{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60f9091c-4e70-467b-b999-fe6110f3d170",
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import Markdown, HTML\n",
    "import malariagen_data\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pyprojroot import here\n",
    "import yaml\n",
    "import dask\n",
    "dask.config.set(scheduler=dask_scheduler);\n",
    "from textwrap import dedent\n",
    "import geopandas as gpd\n",
    "import bokeh.layouts as bklay\n",
    "import bokeh.plotting as bkplt\n",
    "import bokeh.models as bkmod\n",
    "from bokeh.io import output_notebook\n",
    "import plotly.express as px\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43980486-70ba-425d-b95f-0a0ab7273038",
   "metadata": {},
   "outputs": [],
   "source": [
    "ag3 = malariagen_data.Ag3(\n",
    "    # pin the version of the cohorts analysis for reproducibility\n",
    "    cohorts_analysis=cohorts_analysis,\n",
    "    results_cache=(here() / \"build\" / \"malariagen_data_cache\").as_posix(),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0f8f708-3a18-4419-85b8-9dbd704bfa5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf_cohorts = gpd.read_file(\n",
    "    here() / \n",
    "    \"build\" / \n",
    "    analysis_version / \n",
    "    \"analysis\" / \n",
    "    \"final_cohorts.geojson\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8166da36-e789-43b0-afd1-9dc2f94e0ecb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def stack_overlaps(df, start_col, end_col, tolerance=10000):\n",
    "    occupants = [None]\n",
    "    out = []\n",
    "    for _, cur in df.iterrows():\n",
    "\n",
    "        level = 0\n",
    "        prv = occupants[level]\n",
    "        # Search upwards to find the first vacant level.\n",
    "        while prv is not None and cur[start_col] <= (prv[end_col] + tolerance):\n",
    "            level += 1\n",
    "            if level == len(occupants):\n",
    "                occupants.append(None)\n",
    "            prv = occupants[level]\n",
    "        occupants[level] = cur\n",
    "        out.append(level)\n",
    "    return np.asarray(out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c0732e1-7b49-45f2-9f3f-e390aaf67b81",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_cohort_signals(contig, cohort_id):\n",
    "    try:\n",
    "        df_signals = pd.read_csv(\n",
    "            here()\n",
    "            / \"build\"\n",
    "            / analysis_version\n",
    "            / \"analysis\"\n",
    "            / \"h12-signal-detection\"\n",
    "            / f\"{cohort_id}_{contig}.csv\"\n",
    "        )\n",
    "    except pd.errors.EmptyDataError:\n",
    "        df_signals = pd.DataFrame()\n",
    "    return df_signals\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "671c9b60-e55c-419b-b28e-a7df742dc076",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_signals(contig, start=None, stop=None):\n",
    "    \n",
    "    # Load signal dataframes for all cohorts.\n",
    "    dfs = []\n",
    "    for _, row in gdf_cohorts.iterrows():\n",
    "        df = load_cohort_signals(contig=contig, cohort_id=row['cohort_id'])\n",
    "        dfs.append(df)\n",
    "    df_signals = pd.concat(dfs, axis=0).assign(statistic=\"H12\")\n",
    "\n",
    "    # Merge with cohorts data.\n",
    "    df_signals = df_signals.merge(gdf_cohorts, on=\"cohort_id\")\n",
    "    \n",
    "    # Color by taxon.\n",
    "    color_dict = {\n",
    "        'gambiae': '#BEC4FF',\n",
    "        'coluzzii': '#D7B2A6',\n",
    "        'arabiensis': '#A6D7CA',\n",
    "    }\n",
    "    df_signals['color'] = df_signals['taxon'].map(color_dict).fillna('lightgrey')\n",
    "    \n",
    "    # Filter to region.\n",
    "    if start and stop:\n",
    "        df_signals = df_signals.query(\n",
    "        f\"focus_pstop < {int(stop)} and focus_pstart > {int(start)}\"\n",
    "    )\n",
    "        \n",
    "    # Sort and stack\n",
    "    df_signals = df_signals.sort_values(by='span2_pstart')\n",
    "    df_signals['level'] = stack_overlaps(df_signals, 'span2_pstart', 'span2_pstop')\n",
    "    \n",
    "    return df_signals\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f438638c-821a-4f9a-bed5-8ff84174d394",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_signals(\n",
    "    df, \n",
    "    contig, \n",
    "    patch_height=0.7, \n",
    "    row_height=20,\n",
    "    genes_height=80,\n",
    "    x_min=None, \n",
    "    x_max=None,\n",
    "):\n",
    "    \n",
    "    # Default to plotting the whole contig.\n",
    "    if x_min is None:\n",
    "        x_min = 0\n",
    "    if x_max is None:\n",
    "        x_max = ag3.genome_sequence(contig).shape[0]\n",
    "\n",
    "    # Set up triangle shapes for bokeh patches glyphs.\n",
    "    left_xs = [\n",
    "        np.array([row.span2_pstart, row.focus_pstart, row.focus_pstart])\n",
    "        for idx, row in df.iterrows()\n",
    "    ]\n",
    "    left_ys = [\n",
    "        np.array([row.level + patch_height / 2, row.level, row.level + patch_height])\n",
    "        for idx, row in df.iterrows()\n",
    "    ]\n",
    "    right_xs = [\n",
    "        np.array([row.focus_pstop, row.focus_pstop, row.span2_pstop])\n",
    "        for idx, row in df.iterrows()\n",
    "    ]\n",
    "    right_ys = [\n",
    "        np.array([row.level, row.level + patch_height, row.level + patch_height / 2])\n",
    "        for idx, row in df.iterrows()\n",
    "    ]\n",
    "    center_xs = [\n",
    "        np.array([row.pcenter, row.pcenter])\n",
    "        for idx, row in df.iterrows()\n",
    "    ]\n",
    "    center_ys = [\n",
    "        np.array([row.level, row.level + patch_height])\n",
    "        for idx, row in df.iterrows()\n",
    "    ]\n",
    "\n",
    "    source = bkmod.ColumnDataSource(data={\n",
    "        'cohort': df.cohort_id,\n",
    "        'taxon': df.taxon,\n",
    "        'statistic': df.statistic,\n",
    "        'chromosome': df.contig,\n",
    "        'score': df.delta_i.astype(int),\n",
    "        'peak_start': df.span2_pstart,\n",
    "        'peak_stop': df.span2_pstop,\n",
    "        'focus_start': df.focus_pstart,\n",
    "        'focus_stop': df.focus_pstop,    \n",
    "        'left_xs':left_xs,\n",
    "        'left_ys':left_ys,\n",
    "        'right_xs':right_xs,\n",
    "        'right_ys':right_ys,\n",
    "        'center_xs': center_xs,\n",
    "        'center_ys': center_ys,\n",
    "        'bottom': df.level,\n",
    "        'mid': df.level + .5,\n",
    "        'top': df.level + patch_height,\n",
    "        'taxon_color':df.color\n",
    "    })\n",
    "\n",
    "    hover = bkmod.HoverTool(\n",
    "        tooltips=[\n",
    "            (\"Cohort\", '@cohort'),\n",
    "            (\"Statistic\", '@statistic'),\n",
    "            (\"Score\", '@score'),\n",
    "            (\"Focus\", \"@focus_start{,} - @focus_stop{,}\"),\n",
    "        ],\n",
    "    )\n",
    "\n",
    "    xwheel_zoom = bkmod.WheelZoomTool(\n",
    "        dimensions=\"width\", maintain_focus=False\n",
    "    )\n",
    "    \n",
    "    # make figure \n",
    "    fig1 = bkplt.figure(\n",
    "        title='Selection signals',\n",
    "        width=900, \n",
    "        height=80 + (row_height * max(df.level)), \n",
    "        tools=[\"tap\" ,\"xpan\", \"xzoom_in\", \"xzoom_out\",\"reset\", xwheel_zoom, hover],\n",
    "        toolbar_location='above', \n",
    "        active_drag='xpan', \n",
    "        x_range = bkmod.Range1d(x_min, x_max, bounds='auto'),\n",
    "        y_range = bkmod.Range1d(-0.5, max(df.level) + 1.3, bounds='auto'),\n",
    "        active_scroll=xwheel_zoom,\n",
    "    )\n",
    "\n",
    "    fig1.patches(\n",
    "        xs='left_xs', \n",
    "        ys='left_ys', \n",
    "        source=source, \n",
    "        color=\"taxon_color\", \n",
    "        alpha=.7, \n",
    "        line_width=2, \n",
    "        # legend_field='taxon',\n",
    "    )\n",
    "    fig1.patches(\n",
    "        xs='right_xs', \n",
    "        ys='right_ys', \n",
    "        source=source, \n",
    "        color=\"taxon_color\", \n",
    "        alpha=.7, \n",
    "        line_width=2, \n",
    "        # legend_field='taxon',\n",
    "    )\n",
    "\n",
    "    fig1.quad(\n",
    "        bottom='bottom', \n",
    "        top='top', \n",
    "        left='focus_start', \n",
    "        right='focus_stop', \n",
    "        source=source, \n",
    "        color=\"red\", \n",
    "        alpha=.5, \n",
    "        line_width=2,\n",
    "    )\n",
    "\n",
    "    glyph = bkmod.MultiLine(\n",
    "        xs='center_xs', \n",
    "        ys='center_ys', \n",
    "        line_color='red', \n",
    "        line_width=2, \n",
    "        line_alpha=0.8,\n",
    "    )\n",
    "    fig1.add_glyph(source, glyph)\n",
    "    \n",
    "    # tidy up the plot \n",
    "    fig1.yaxis.visible = False\n",
    "    fig1.xaxis.visible = False\n",
    "    fig1.ygrid.visible = False\n",
    "    # fig1.legend.background_fill_alpha = 0.2\n",
    "\n",
    "    url = '../cohort/@cohort.html'\n",
    "    taptool = fig1.select(type=bkmod.TapTool)\n",
    "    taptool.callback = bkmod.OpenURL(url=url)\n",
    "\n",
    "    fig2 = ag3.plot_genes(\n",
    "        region=contig, \n",
    "        sizing_mode=\"stretch_width\",\n",
    "        x_range=fig1.x_range,\n",
    "        height=genes_height,\n",
    "        show=False,\n",
    "    )\n",
    "    \n",
    "    fig = bklay.gridplot(\n",
    "        [fig1, fig2],\n",
    "        ncols=1,\n",
    "        toolbar_location=\"above\",\n",
    "        merge_tools=True,\n",
    "        sizing_mode=\"stretch_width\",\n",
    "    ) \n",
    "\n",
    "    bkplt.show(fig)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab6be74f-6d6c-4440-96f4-800f44f14f67",
   "metadata": {},
   "outputs": [],
   "source": [
    "h12_calibration_dir = f\"build/{analysis_version}/analysis/h12-calibration\"\n",
    "g123_calibration_dir = f\"build/{analysis_version}/analysis/g123-calibration\"\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "selection-atlas-2025-02-06",
   "language": "python",
   "name": "selection-atlas-2025-02-06"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
